libraries/stdlib/jvm/src/kotlin/collections/ArraysJVM.kt

```diff
@@ -35,7 +35,7 @@ public inline fun ByteArray.toString(charset: Charset): String = String(this, ch
 public actual inline fun <reified T> Collection<T>.toTypedArray(): Array<T> {
     @Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
     val thisCollection = this as java.util.Collection<T>
-    return thisCollection.toArray(arrayOfNulls<T>(0)) as Array<T>
+    return thisCollection.toArray(arrayOfNulls<T>(thisCollection.size())) as Array<T>
```

Comment:
There are a few reasons to avoid doing that:
- `size` is not guaranteed to be constant-time operation and if `toArray` needs to check the size anyway, we'll end up scanning the collection twice;
- the collection may change concurrently (👋 from `ConcurrentSkipList` ;)), so pre-allocated array may not fit the data and will be re-allocated;
- and finally, the main reason: pre-allocating the array will be slower on Hotspot JVM (see https://shipilev.net/blog/2016/arrays-wisdom-ancients/; tl;dr: it's all about intrinsics and array's zeroing; that's still true for JDK 24).

On the other hand, there's no need to allocate an empty array every single time, so it would be better to introduce an EMPTY_ARRAY constant and pass it to `toArray`.

Reply:
A shared empty array can't work  unfortunately because the resulting array won't have the correct component type.

Reply:
Right hence my attempt. But yeah non-constant size queries and concurrency are issues indeed :/